webknossos-datastore/app/com/scalableminds/webknossos/datastore/tracings/skeleton/TreeValidator.scala

```diff
@@ -102,7 +110,51 @@ object TreeValidator {
       }
     }
 
+  private def checkNoDuplicateTreeGroupIds(treeGroups: Seq[TreeGroup]) = {
+    val treeGroupIds = getAllTreeGroupIds(treeGroups, Seq[Int]())
+    val distinctTreeGroupIds = treeGroupIds.distinct
+    if (treeGroupIds.size == distinctTreeGroupIds.size) {
+      Full(())
+    } else {
+      Failure(s"Duplicate treeGroupIds: ${treeGroupIds.diff(distinctTreeGroupIds).mkString(", ")}")
+    }
+  }
+
+  private def checkAllNodesUsedInCommentsExist(trees: Seq[Tree], comments: Seq[Comment]): Box[Unit] = {
+    val nodesInAllTrees = trees.flatMap(_.nodes.map(_.id))
+    val nodesInComments = comments.map(_.nodeId).distinct
+
+    val nodesOnlyInComments = nodesInComments.diff(nodesInAllTrees)
+    if (nodesOnlyInComments.isEmpty) {
+      Full(())
+    } else {
+      Failure(s"Some comments refer to non-existent nodes. comments: ${nodesOnlyInComments.mkString(", ")}")
+    }
+  }
+
+  private def checkAllNodesUsedInBranchPointsExist(trees: Seq[Tree], branchPoints: Seq[BranchPoint]): Box[Unit] = {
```

Comment:
Please make this and similar functions a bit more DRY. There are multiple functions checking that some nodes exist in the trees.